### PR TITLE
chore: DP#7155 Update analytics Part 1

### PR DESCRIPTION
**Pull request by tihuan**
⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️   CHECK LIST BEFORE MERGE ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ 
- [x] Update Plausible events
- [x] Remove console.log after Amanda verifies the events

<details open>
  <summary>Argus Stack Details</summary>
  <br />
  <table>
    <tr>
      <th>Name</th>
      <td>thuang-update-analytics</td>
    </tr>
    <tr>
      <th>Short Name</th>
      <td>hardy-civet</td>
    </tr>
    <tr>
      <th>Base URL</th>
      <td><a href="https://hardy-civet.dev-sc.dev.czi.team" target="_blank">https://hardy-civet.dev-sc.dev.czi.team</a></td>
    </tr>
  </table>
</details>

---

1. This PR addresses 9 analytics event changes - please use the [rdev](https://hardy-civet.dev-sc.dev.czi.team/d/super-cool-spatial.cxg/?spatial=true&dl=true) to test the affected analytics events 🙏 
2. Introduce more thunk actions, since many new events require reading the latest Redux state to determine whether an event should be triggered or not
3. In `client/src/analytics/index.ts` I added more analytics functions and prefix the thunk ones accordingly

https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/gh/chanzuckerberg/single-cell-data-portal/7155

Screenshot 2024-07-18 at 8 08 18 AM

### CODE REVIEW

**Review by tihuan — COMMENTED** [review]
Actually let me refactor a bit!

File: client/src/components/categorical/category/index.tsx
Comment (tihuan) [inline client/src/components/categorical/category/index.tsx:121]: we need `isColorAccessor` in order to know whether the user is toggling on or off the color mode for analytics. We only want to trigger the event when the user is turning ON the color mode

File: client/src/components/categorical/index.tsx
Comment (tihuan) [inline client/src/components/categorical/index.tsx:37]: refactor to use the common `connect(mapStateToProps)(Categories);` pattern, so we're one step closer to transition from class to functional component

File: client/src/components/categorical/index.tsx
Comment (tihuan) [inline client/src/components/categorical/index.tsx:35]: this is now a Redux state since thunk analytics actions need it. Also array is more Redux friendly than Set, since array is easier to copy and serialize without extra steps

**Comment by tihuan** [comment]
Hey @ainfeld ! The events included in the PR descriptions are ready for your verification now 🙏 

Please use this rdev:
https://hardy-civet.dev-sc.dev.czi.team/d/super-cool-spatial.cxg/?spatial=true&dl=true

Note that you'll need to open the dev console in your browser to look at the logs, since rdev is not connected to Plausible!

Thank you!

Screenshot 2024-07-18 at 10 55 17 AM

**Comment by tihuan** [comment]
Addressing Amanda's feedback!

1. `EXPLORER_COLORBY_CATEGORIES_BUTTON_CLICKED`  and `EXPLORER_COLORBY_HISTOGRAM_CONTINUOUS_BUTTON_CLICKED` -> fire when color by is turned on AND off -> don't send events when OFF ✅

1. `EXPLORER_SELECT_HISTOGRAM` -> fires when histogram is selected and deselected -> don't send events when DESELECTED ✅

https://github.com/user-attachments/assets/407baa7f-ef55-4708-b966-a066486e1dfa

File: client/src/components/categorical/category/index.tsx
Comment (rainandbare) [inline client/src/components/categorical/category/index.tsx:511]: Slowly but surely!

File: client/src/components/embedding/index.tsx
Comment (rainandbare) [inline client/src/components/embedding/index.tsx:103]: ♥️

File: client/src/components/categorical/category/index.tsx
Comment (tihuan) [reply client/src/components/categorical/category/index.tsx]: Oh actually it's a functional component! I just went ahead and `useCallback()` on it 😁 Thanks for catching this!

File: client/src/analytics/index.ts
Comment (tihuan) [reply client/src/analytics/index.ts]: Just commented it out 😆 Thank you!

File: client/src/components/categorical/category/index.tsx
Comment (rainandbare) [reply client/src/components/categorical/category/index.tsx]: Nice!

**Review by rainandbare — APPROVED** [review]
LGTM! Might want another pair of eyes that is more familiar with the code base though.

**Comment by tihuan** [comment]
Thanks so much for the thorough review, @rainandbare 🤩 Will wait for @kaloster 's review too 👍

File: client/src/components/categorical/category/index.tsx
Comment (kaloster) [inline client/src/components/categorical/category/index.tsx:129]: If `isColorAccessor` is `true`, `!isColorAccessor` will be `false`

File: client/src/components/categorical/index.tsx
Comment (kaloster) [reply client/src/components/categorical/index.tsx:35]: Array is slightly slower for larger arrays and Set is more performant, but I think for our array size that should be negligible?

File: client/src/components/categorical/index.tsx
Comment (kaloster) [inline client/src/components/categorical/index.tsx:196]: super nit - I started putting these in the types file to make the component even smaller

**Review by kaloster — COMMENTED** [review]
Just one comment on the `isColorAccessor` and a couple of nits.

Thanks for the great update and cleanup!

File: client/src/components/categorical/category/index.tsx
Comment (tihuan) [reply client/src/components/categorical/category/index.tsx]: Yeah I'm struggling to make this clearer lol So the idea is that `handleColorChange()` will be called with the category's current `isColorAccessor` state. So if it's `true`, it means that the category is already being used to color the graph, so we will be turning it off. So if it's `false`, it means that the user is about to color the graph using this category, thus us tracking the event!

I just updated the argument name from `isColorAccessor` to `currentIsColorAccessor` - not sure if that helps or not 😆

File: client/src/components/categorical/index.tsx
Comment (tihuan) [reply client/src/components/categorical/index.tsx:196]: Oh that's right! Let me extract `types.ts` file too thank you!

File: client/src/components/categorical/category/index.tsx
Comment (kaloster) [reply client/src/components/categorical/category/index.tsx]: OK I see now. I think I was confused of by the inline comment  lol

> If we're going from `isColorAccessor` being `false` to `true`,
     * we should track the event!

**Review by kaloster — APPROVED** [review]

File: client/src/components/categorical/types.ts
Comment (tihuan) [inline client/src/components/categorical/types.ts:1]: @kaloster extracted types out in its own file!

**Pull request merged**
